### ISSUE TITLE
GSoC and the alternative routes

### ISSUE THREAD

**Issue by melonSkin76** [organicmaps/organicmaps#2237]
Hi everyone!

I am Runqiu (you can also call me Thomas), a CS student at the University of Minnesota. Recently I'm planning to take part in the GSoC and I am very interested in your project. More specifically, according to the [ideas list](https://github.com/organicmaps/organicmaps/wiki/GSoC-2022-ideas), I'm interested in the idea of finding alternative routes. But I really want to understand the current routing logic. Could you please give me some advice on where I can start? For example, is the routing process supposed to be online or offline? As far as I am concerned, there's plenty of online routing APIs we can use for OSM, e.g. OSRM, GraphHopper, etc. Or we can implement some routing algorithms by ourselves, which is much harder.

Thank you!

**Comment by biodranik** [comment]
There is already an offline routing algorithm (modified bidirectional A*) implemented in our code. The idea is to use different criteria in it and produce different alternate routes, ideally at the same time, to avoid it running several times.

**Comment by arnaudvergnet** [comment]
@biodranik this would be a great opportunity to write some documentation on the routing algorithm :)

**Comment by biodranik** [comment]
Agree. But the basic principle of the A* is [simple](https://en.wikipedia.org/wiki/A*_search_algorithm). We run it from both start and end points to make it faster.

**Comment by biodranik** [comment]
See also #2153

**Comment by strump** [comment]
Link to PR #3447
